packages/logger/src/formatters/ConsoleFormatter.ts

```diff
@@ -0,0 +1,77 @@
+import chalk from 'chalk';
+import {info, warning, error} from 'log-symbols';
+import prettyMs from 'pretty-ms';
+
+import {LogLevel, FormatEntry, Formatter} from '../types';
+
+export class ConsoleFormatter implements Formatter {
+  private startTime: [number, number];
+
+  constructor() {
+    this.startTime = process.hrtime();
+  }
+
+  next(entry: FormatEntry) {
+    const timeDiff = process.hrtime(this.startTime);
+    const msDiff = timeDiff[0] * 1000 + timeDiff[1] / 1000;
+    const timestamp = chalk.magenta(`+${prettyMs(msDiff)}`);
+
+    const {level, payload, scopes} = entry;
+
+    let logMsg = `${payload} ${timestamp}`;
+
+    if (level === LogLevel.Critical) {
+      const {message, stack} = payload as Error;
+      logMsg = message;
+
+      if (stack != null) {
+        const [name, ...lines] = stack.split('\n');
+        logMsg = `${name} ${timestamp}\n${lines
+          .map(line => {
+            return chalk.gray(line);
+          })
+          .join('\n')}`;
+      }
+    }
+
+    const prefix = `[${scopes.join(':')}] `;
+    const {logFn, icon, chalkColor} = configForLogLevel(level);
+    logFn(
+      prefix,
+      icon,
+      spaces(1),
+      chalkColor.bold.underline(level),
+      spaces(8 - level.length),
+      logMsg,
+    );
+  }
+}
+
+function spaces(num: number) {
+  return (spaces as any).memoize[num] || new Array(num).fill(' ').join('');
+}
+(spaces as any).memoize = {};
```

Comment:
A few things we can change here:
- Use [`string.repeat()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/repeat) over this array stuff
- I don't think this memoization is working right now
- We can kill the `any` and `as` pretty easily by using `Map`

```typescript
const spacesCache = new Map<number, string>();

function spaces(number: number) {
  const cachedString = spacesCache.get(number);

  if (cachedString) {
    return cachedString;
  }

  const newString = ' '.repeat(number);
  spacesCache.set(number, newString);
  return newString;
}
```

Reply:
though tbh I am not sure this isnt a premature optimization